packages/network-controller/src/NetworkController.ts

```diff
@@ -63,32 +73,178 @@ export type NetworkMetadata = {
 };
 
 /**
- * Custom RPC network information
+ * The type of an RPC endpoint.
+ *
+ * @see CustomRpcEndpoint
+ * @see InfuraRpcEndpoint
+ */
+export enum RpcEndpointType {
+  Custom = 'custom',
+  Infura = 'infura',
+}
+
+/**
+ * An Infura RPC endpoint is a reference to a specific network that Infura
+ * supports as well as an Infura account we own that we allow users to make use
+ * of for free. We need to disambiguate these endpoints from custom RPC
+ * endpoints, because while the types for these kinds of object both have the
+ * same interface, the URL for an Infura endpoint contains the Infura project
+ * ID, and we don't want this to be present in state. We therefore hide it by
+ * representing it in the URL as `{infuraProjectId}`, which we replace this when
+ * create network clients. But we need to know somehow that we only need to do
+ * this replacement for Infura endpoints and not custom endpoints — hence the
+ * type.
+ */
+export type InfuraRpcEndpoint = {
+  /**
+   * The user-facing name of the endpoint.
+   */
+  name: string;
+  /**
+   * The identifier for the network client that has been created for this RPC
+   * endpoint.
+   */
+  networkClientId: InfuraNetworkType;
+  /**
+   * The type of this endpoint, always "default".
+   */
+  type: RpcEndpointType.Infura;
+  /**
+   * The URL of the endpoint. Expected to be a sort of template with the text
+   * `{infuraProjectId}`, which will get replaced with the Infura project ID.
+   */
+  url: string;
+};
+
+/**
+ * A custom RPC endpoint is a reference to a user-defined server which fronts an
+ * EVM chain. It may refer to an Infura network, but only by coincidence.
+ */
+export type CustomRpcEndpoint = {
+  /**
+   * The user-facing name of the endpoint.
+   */
+  name: string;
+  /**
+   * The identifier for the network client that has been created for this RPC
+   * endpoint.
+   */
+  networkClientId: string;
+  /**
+   * The type of this endpoint, always "custom".
+   */
+  type: RpcEndpointType.Custom;
+  /**
+   * The URL of the endpoint.
+   */
+  url: string;
+};
+
+/**
+ * An RPC endpoint is a reference to a server which fronts an EVM chain. There
+ * are two varieties of RPC endpoints: Infura and custom.
+ *
+ * @see CustomRpcEndpoint
+ * @see InfuraRpcEndpoint
+ */
```

Comment:
nit:
```suggestion
 * @see {@link CustomRpcEndpoint}
 * @see {@link InfuraRpcEndpoint}
 */
```

Reply:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).